### ISSUE TITLE
Error using parent_role for pundit authorization

### ISSUE THREAD

**Issue by thebadmonkeydev** [rmosolgo/graphql-ruby#770]
| package | version |
|--|--|
| graphql-ruby | 1.5.13 |
| graphql-pro | 1.4.2 |
| rails | 5.0.1 |

I'm looking to add compile-time authorization to the `transfer` field on my `TransactionType` like below:

```ruby
Types::TransactionType = GraphQL::ObjectType.define do
  name 'Transaction'

  field :transfer, Types::TransferTransactionType do
    access(parent_role: :store_admin)

    resolve -> (*) {
      # basic resolver stuff
    }   
  end 
end
```

We have an `app/policies/transaction_policy.rb` that has a method `store_admin?` on it.  When I use the above like is mentioned in the docs [here](http://graphql-ruby.org/pro/authorization.html#authorize-values-by-parent), I get `NoMethodError: undefined method 'new' for nil:NilClass`:

```
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-pro-1.4.2/lib/graphql/pro/access/analyzer.rb:65:in `allowed_gate?'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-pro-1.4.2/lib/graphql/pro/access/analyzer.rb:46:in `allowed_field?'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-pro-1.4.2/lib/graphql/pro/access/analyzer.rb:23:in `call'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/reducer_state.rb:15:in `call'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:47:in `block in visit_analyzers'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:46:in `each'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:46:in `visit_analyzers'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:34:in `reduce_node'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:38:in `block (2 levels) in reduce_node'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:37:in `each'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:37:in `block in reduce_node'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:36:in `each'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:36:in `reduce_node'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:38:in `block (2 levels) in reduce_node'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:37:in `each'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:37:in `block in reduce_node'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:36:in `each'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:36:in `reduce_node'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:22:in `block in analyze_query'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:21:in `each'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/analysis/analyze_query.rb:21:in `analyze_query'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/query/validation_pipeline.rb:85:in `ensure_has_validated'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/query/validation_pipeline.rb:33:in `valid?'
/home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.5.13/lib/graphql/query.rb:122:in `result'
```

The query that we're using is basically the following with most
selections removed for brevity:

```
{
  transaction(id: 123, type: "Order") {
    transfer {
      id  
    }   
  }
}
```

I've tried explicitly setting the parent pundit policy name like this:

```
access(parent_role: :store_admin, parent_pundit_policy_name: 'TransactionPolicy')
```

and receive the same error with the same backtrace.

**Comment by rmosolgo** [comment]
Hi, thanks for the great bug report! I'm looking into a couple ideas, including raising an error sooner when graphql-pro fails to find the corresponding policy (right now the error is rescued silently, even when an explicit name is provided 😖 )

In the mean time, could you check the stack trace one more time? I think there might be _one_ more entry, perhaps on the same line as `NoMethodError: undefined method 'new'` (I expect it to point to `pundit_strategy.rb`). This will give me a bit better understanding of just what went wrong. If you see it, could you add it here?

**Comment by thebadmonkeydev** [comment]
You're absolutely right! I missed the first line of the backtrace.  The full backtrace as it's printed by rspec is below:

```
NoMethodError:
  undefined method `new' for nil:NilClass
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-pro-1.4.2/lib/graphql/pro/access/pundit_strategy.rb:16:in `allowed?'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-pro-1.4.2/lib/graphql/pro/access/analyzer.rb:65:in `allowed_gate?'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-pro-1.4.2/lib/graphql/pro/access/analyzer.rb:46:in `allowed_field?'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-pro-1.4.2/lib/graphql/pro/access/analyzer.rb:23:in `call'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/reducer_state.rb:15:in `call'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:68:in `block in visit_analyzers'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:67:in `each'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:67:in `visit_analyzers'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:55:in `reduce_node'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:59:in `block (2 levels) in reduce_node'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:58:in `each'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:58:in `block in reduce_node'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:57:in `each'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:57:in `reduce_node'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:59:in `block (2 levels) in reduce_node'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:58:in `each'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:58:in `block in reduce_node'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:57:in `each'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:57:in `reduce_node'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:43:in `block in analyze_query'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:42:in `each'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:42:in `analyze_query'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:11:in `block in analyze_multiplex'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:9:in `map'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/analysis/analyze_query.rb:9:in `analyze_multiplex'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/execution/multiplex.rb:79:in `run_as_multiplex'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/execution/multiplex.rb:57:in `run_queries'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/query.rb:115:in `block in result'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/query.rb:286:in `with_prepared_ast'
# /home/michael/.rvm/gems/ruby-2.3.1/gems/graphql-1.6.2/lib/graphql/query.rb:114:in `result'
# ./spec/graphql/queries/transfer_on_transaction_queries_spec.rb:43:in `block (2 levels) in <top (required)>'
# ./spec/graphql/queries/transfer_on_transaction_queries_spec.rb:81:in `block (2 levels) in <top (required)>'
```

Sorry about that :man_facepalming:

**Comment by rmosolgo** [comment]
🤔 🤔 🤔  Thanks for posting that, but I still can't figure it out! 😖 

The gem includes a build-time check to make sure all policies are found, but for some reason, the check _isn't_ working in this case.  I'm trying to add test cases to my test to find this one, but I can't nail it down!

Could you try running that test again, but put this monkeypatch in the test setup, and then share the output?

```ruby 
module GraphQL::Pro::Access::Instrumentation::AssertPunditPolicy
  class << self
    alias :original_assert_policy :assert_policy

    # Print debug output for types whose names start with: 
    TYPE_FILTER = "Transaction"

    # Wrap the built-in policy check with some debug logging:
    def assert_policy(owner, field, gate)
      if gate && owner.is_a?(GraphQL::BaseType) && (TYPE_FILTER.nil? || owner.name.starts_with?(TYPE_FILTER))
        print "Pundit #{gate.level}: #{owner}#{field ? ".#{field.name}" : "" }"
        original_assert_policy(owner, field, gate)
        puts " -> OK (#{gate.pundit_policy.inspect}, #{gate.parent_gate ? gate.parent_gate.pundit_policy.inspect : "nil"})"
      else
        original_assert_policy(owner, field, gate)
      end
    rescue GraphQL::Pro::Access::PunditStrategyNotFoundError => err
      puts "\n-> FAIL #{err}"
      raise err
    end
  end
end
```

That will show us if the check is running _at all_, because it should output 

```
Pundit access: Transaction.transfer -> OK (nil, TransactionPolicy)
```

But if it outputs _nothing_, then we know that for some reason, `Transaction` isn't being checked ahead of time. 

Also, could you share your usage of `authorize(...)` in your schema definition, and your query execution code from `transfer_on_transaction_queries_spec.rb:43`, including the Ruby code for running the query and the minimal query string which reproduces this error? (I saw the simplified query string in your first comment, does that query string reproduce this error?)

Sorry for the hassle on this but thanks for working with me!!

**Comment by thebadmonkeydev** [comment]
No worries, @rmosolgo makes me feel like I'm helping :smiley: 

I ran the spec with the monkey patch like you asked and it did output similarly to what you described:

```
Pundit access: Transaction.transfer -> OK (nil, TransactionPolicy)
Pundit access: Transaction.member_type -> OK (TransactionPolicy, nil)
```

There's a second one there because another field on that type is authorized as well and my query in that case selected both.

Our schema looks like this:

```ruby
ArcadeSchema = GraphQL::Schema.define do
  monitoring    :new_relic
  authorization(:pundit, current_user: :member)

  raise_definition_error true

  query    Types::QueryType
  mutation Types::MutationType

  resolve_type -> (obj, _) do
    "Types::#{obj.class.name}Type".constantize
  end
end
```

The query I posted above does reproduce the error in my case.  It is the selection of the `transfer` field on `Types::TransactionType` that causes the failure.

Regarding my actual query approach...we have a method in our `spec_helper.rb` file named `build_query`:

```ruby
def build_query(query, opt={})
  GraphQL::Query.new(ArcadeSchema, query, opt)
end
```
The in individual tests we'll do something like:

```ruby
let(:query) do
  %Q{
    {
      transaction(#{args}) {
        transfer {
          id
          initiated_at
          payments_fee
          platform_fee
          payments_fee_percent
          platform_fee_percent
          amount
        }
      }
    }
  }
end
subject { build_query(query, context: { member: user }) }
```
and test against `subject.result`. _Above is the actual query used in the spec file, BTW_

**Comment by thebadmonkeydev** [comment]
Sorry....in the above `let(:query)` block, `args` is just a string that looks like this:

```ruby
"id: #{order.id} type: \"Order\""
```
it's separate so that I can easily spec against different parameters without having to repeat the entire query.

**Comment by rmosolgo** [comment]
Ok, I took a look with fresh eyes and now it makes sense! 

`parent_role` is supported for `authorize` (runtime check), but not `access` (static check). Here's why: 

`parent_role` was designed to show certain attributes for valid viewer-object combinations. For example, a viewer can see a patient's medical records _only_ if the viewer is assigned to that patient's case. 

In order to make this check, you must have access to the _specific_ patient (the runtime object "behind" the GraphQL response), so it's only possible to make this check at runtime. 

During static auth checks, we don't have access to specific objects, only their GraphQL types. So, we can't make assertions based on parent objects. We don't have any objects yet because we haven't started running the query. 

How does that sound to you? 

To put it another way,  How would these two differ? 

```ruby 
  # Static check on parent object (not possible)
  field :transfer, Types::TransferTransactionType,  access: { parent_role: :store_admin }
  # OR 
  # Runtime check on parent object
  field :transfer, Types::TransferTransactionType,  authorize: { parent_role: :store_admin }
```

Would `authorize:` work for you? Or, can you describe your goal, and we can see if there's something to make it work? 

Sorry for the run-around on this one! If `authorize` works for you, I'll update the gem to give a better warning in this case 😬

**Comment by thebadmonkeydev** [comment]
Thanks for the detailed explanation @rmosolgo and sorry for the delay responding to you...My main reason for wanting to use `access` over `authorize` in this case is that I want the query to fail with an error instead of just responding with a `nil` value for that field.  Is it possible to configure the action taken on different authorization failures?

**Comment by rmosolgo** [comment]
No, that wasn't possible, but you're right that it's a great feature! 

So I just pushed version 1.4.4 which includes a way to provide your own behavior. You can see the docs here: http://graphql-ruby.org/pro/authorization#unauthorized-object

How does that look to you??

**Comment by thebadmonkeydev** [comment]
That looks perfect @rmosolgo !  Sorry for the delayed response.  I'll get us updated to 1.4.4 and try everything out.  But I'm pretty sure you can close this issue as this resolves my need for `parent_role` and whatnot.
